### PR TITLE
feat: integrate deepgemm into EPMoE

### PR DESCRIPTION
**Pull request by TianQiLin666666**
## Motivation

For normal EPMoE (no DeepEP),  integrate DeepGEMM as an option. 

## Modifications

1. Add `forward_deepgemm` in EPMoE.  Use env `EPMOE_USE_DEEPGEMM` to enable it.
2. Add some triton kernels for PreRecord and PostRecord of `forward_deepgemm`.

### Speed
With 2*H20-96G*8 for EP16, enabling EPMOE_USE_DEEPGEMM leads to a 14% throughput gain.
* Disable EPMOE_USE_DEEPGEMM
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max reqeuest concurrency:                64
Successful requests:                     192
Benchmark duration (s):                  280.53
Total input tokens:                      672000
Total generated tokens:                  288000
Total generated tokens (retokenized):    285278
Request throughput (req/s):              0.68
Input token throughput (tok/s):          2395.46
Output token throughput (tok/s):         1026.63
Total token throughput (tok/s):          3422.09
Concurrency:                             53.17
Accept length:                           3.04
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   77682.09
Median E2E Latency (ms):                 79075.32
---------------Time to First Token----------------
Mean TTFT (ms):                          4389.81
Median TTFT (ms):                        758.88
P99 TTFT (ms):                           21329.72
---------------Inter-Token Latency----------------
Mean ITL (ms):                           48.90
Median ITL (ms):                         34.40
P95 ITL (ms):                            121.80
P99 ITL (ms):                            264.88
Max ITL (ms):                            21210.85
==================================================
```

* Enable EPMOE_USE_DEEPGEMM
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max reqeuest concurrency:                64
Successful requests:                     192
Benchmark duration (s):                  246.03
Total input tokens:                      672000
Total generated tokens:                  288000
Total generated tokens (retokenized):    285757
Request throughput (req/s):              0.78
Input token throughput (tok/s):          2731.37
Output token throughput (tok/s):         1170.59
Total token throughput (tok/s):          3901.96
Concurrency:                             58.55
Accept length:                           3.10
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   75031.17
Median E2E Latency (ms):                 75272.64
---------------Time to First Token----------------
Mean TTFT (ms):                          4070.91
Median TTFT (ms):                        597.56
P99 TTFT (ms):                           19711.67
---------------Inter-Token Latency----------------
Mean ITL (ms):                           47.35
Median ITL (ms):                         34.07
P95 ITL (ms):                            118.96
P99 ITL (ms):                            244.79
Max ITL (ms):                            19567.36
==================================================
```

* Launch server commands
```
# node0
GLOO_SOCKET_IFNAME=eth0 NCCL_IB_HCA=mlx5_ NCCL_IB_DISABLE=0 NCCL_SOCKET_IFNAME=eth0 NCCL_IB_GID_INDEX=3 \
NCCL_MIN_NCHANNELS=24 \
NCCL_IB_QPS_PER_CONNECTION=8 \
SGL_ENABLE_JIT_DEEPGEMM=1 \
EPMOE_USE_DEEPGEMM=1 \
python3 -m sglang.launch_server \
--cuda-graph-bs 1 2 4 8 10 16 20 24 28 32 40 44 48 52 56 64 66 68 70 72 74 76 --cuda-graph-max-bs 78 \
--attention-backend fa3 \
--speculative-algo NEXTN --speculative-draft /data/models/DeepSeek-R1-NextN --speculative-num-steps 4 --speculative-eagle-topk 2 --speculative-num-draft-tokens 6 \
--model-path /data/models/DeepSeek-R1/ \
--tp 16 \
--dist-init-addr 192.168.0.7:10240 \
--nnodes 2 --node-rank 0 --trust-remote-code \
--host 0.0.0.0 --port 8000 --enable-ep-moe --max-running-requests 78 --mem-fraction-static 0.75 --disable-chunked-prefix-cache

# node1
GLOO_SOCKET_IFNAME=eth0 NCCL_IB_HCA=mlx5_ NCCL_IB_DISABLE=0 NCCL_SOCKET_IFNAME=eth0 NCCL_IB_GID_INDEX=3 \
NCCL_MIN_NCHANNELS=24 \
NCCL_IB_QPS_PER_CONNECTION=8 \
SGL_ENABLE_JIT_DEEPGEMM=1 \
EPMOE_USE_DEEPGEMM=1 \
python3 -m sglang.launch_server \
--cuda-graph-bs 1 2 4 8 10 16 20 24 28 32 40 44 48 52 56 64 66 68 70 72 74 76 --cuda-graph-max-bs 78 \
--attention-backend fa3 \
--speculative-algo NEXTN --speculative-draft /data/models/DeepSeek-R1-NextN --speculative-num-steps 4 --speculative-eagle-topk 2 --speculative-num-draft-tokens 6 \
--model-path /data/models/DeepSeek-R1/ \
--tp 16 \
--dist-init-addr 192.168.0.7:10240 \
--nnodes 2 --node-rank 1 --trust-remote-code \
--host 0.0.0.0 --port 8000 --enable-ep-moe --max-running-requests 78 --mem-fraction-static 0.75 --disable-chunked-prefix-cache

# client
python3 -m sglang.bench_serving --backend sglang \
            --dataset-name random \
            --dataset-path /data/datasets/ShareGPT_V3_unfiltered_cleaned_split.json \
            --random-input-len 3500 \
            --random-output-len 1500 \
            --random-range-ratio 1 \
            --request-rate 74 \
            --max-concurrency 74 \
            --num-prompts 296 \
            --host 0.0.0.0 --port 8000
```

### Accuracy

MMLU test with [mmlu/bench_sglang.py](https://github.com/sgl-project/sglang/blob/main/benchmark/mmlu/bench_sglang.py)
```
100%|██████████████████████████████████████████████████████████████████| 1369/1369 [00:47<00:00, 28.74it/s]
subject: abstract_algebra, #q:100, acc: 0.750
subject: anatomy, #q:135, acc: 0.852
subject: astronomy, #q:152, acc: 0.941
subject: business_ethics, #q:100, acc: 0.880
subject: clinical_knowledge, #q:265, acc: 0.925
subject: college_biology, #q:144, acc: 0.972
subject: college_chemistry, #q:100, acc: 0.630
subject: college_computer_science, #q:100, acc: 0.840
subject: college_mathematics, #q:100, acc: 0.770
subject: college_medicine, #q:173, acc: 0.890
Total latency: 47.643
Average accuracy: 0.865
```

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -47,6 +55,8 @@
 
 logger = logging.getLogger(__name__)
 
+epmoe_use_deepgemm = get_bool_env_var("EPMOE_USE_DEEPGEMM")
```
Comment (xutizhou) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: https://github.com/sgl-project/sglang/blob/b5be56944b6eb61b44866011f157e8df0e563bd7/python/sglang/srt/layers/quantization/deep_gemm.py#L29
We might import it directly.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 2]
```diff
@@ -198,7 +208,130 @@ def __init__(
...
+                else self.w13_weight_scale
+            ),
+        )
+        self.w2_weight_fp8 = (
+            self.w2_weight,
+            self.w2_weight_scale_inv if self.use_block_quant else self.w2_weight_scale,
+        )
+
     def forward(self, hidden_states: torch.Tensor, router_logits: torch.Tensor):
+        if use_deep_gemm and epmoe_use_deepgemm:
```
Comment (xutizhou) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: Why disable EPMOE DeepGEMM when use_deep_gemm is enabled?

Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Maybe forward_deepgemm is called when use_deep_gemm is enabled.

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 1 above)
Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: So, do you mean we just replace EPMOE_USE_DEEPGEMM with _ENABLE_JIT_DEEPGEMM

Comment (xutizhou) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Yes, enabling _ENABLE_JIT_DEEPGEMM will set deepgemm at epmoe as the default configuration.

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 2 above)
Comment (xutizhou) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Are there any cases where Triton GEMM in forward_normal outperforms DeepGEMM?

Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: As for now, I didn't find any case where Triton GEMM in forward_normal outperforms DeepGEMM, but DeepGEMM may occupy more GPU memory.

Comment (xutizhou) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: We could remove epmoe_use_deepgemm and corresponding Environment variable **EPMOE_USE_DEEPGEMM**  for the sake of clarity.

Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: OK, done

**Comment by TianQiLin666666** [comment]
@xutizhou Could you please help me merge this

**Comment by xutizhou** [comment]
> @xutizhou Could you please help me merge this

Sure, I need some time to review and test.

**Comment by ch-wan** [comment]
@TianQiLin666666 Hi! Could you please join our Slack channel or check your Slack message? Thanks.

**Comment by zhyncs** [comment]
Hi @TianQiLin666666 May you help fix the conflicts? Thanks!

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 3]
```diff
@@ -707,3 +707,215 @@ def grouped_gemm_triton(
...
+            scale_dst_ptr = gateup_input_scale_ptr + dst_idx * scale_size
+            for start_offset in tl.range(0, scale_size, BLOCK_SIZE):
+                offset = start_offset + tl.arange(0, BLOCK_SIZE)
+                mask = offset < scale_size
+                in_scale = tl.load(scale_src_ptr + offset, mask=mask)
+                tl.store(scale_dst_ptr + offset, in_scale, mask=mask)
+
+
+def exp2_upper(num: int) -> int:
+    for i in range(2, 31):
```
Comment (xutizhou) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py:784]: why does the variable **num** start from 2**2=4

**Comment by zhyncs** [comment]
Hi @TianQiLin666666 Thanks for the great work! Due to this PR not being updated for a while, and our high interest in this feature, I have asked @xutizhou to make some fixes and optimizations based on your work in a new PR https://github.com/sgl-project/sglang/pull/6821. We will add you as a co-author in the new PR. Thank you for your understanding and help.
